File: Day1-9/7.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"regexp"
	"strconv"
	"strings"
)

type Program struct {
	pid      string
	weight   int
	children []string
}

func parseLine(line string) Program {
	re := regexp.MustCompile("^(.+) \\((\\d+)\\)( -> )?(.+)?")
	matches := re.FindStringSubmatch(line)
	p := Program{}
	p.pid = matches[1]
	p.weight, _ = strconv.Atoi(matches[2])

	if len(matches[4]) > 0 {
		p.children = strings.Split(matches[4], ", ")
	} else {
		p.children = make([]string, 0)
	}

	return p
}

func findSource(programs []Program) Program {
	nonSources := make(map[string]bool)
	for _, p := range programs {
		for _, c := range p.children {
			nonSources[c] = true
		}
	}
	for _, p := range programs {
		if _, ok := nonSources[p.pid]; !ok {
			return p
		}
	}
	panic("Unreachable")
}

func balance(root Program, vertices map[string]Program, weights map[string]int) int {
	for _, c := range root.children {
		fixed := balance(vertices[c], vertices, weights)
		if fixed >= 0 {
			return fixed
		}
	}
	cost := math.MaxInt64
	for _, c := range root.children {
		if weights[c] < cost {
			cost = weights[c]
		}
	}
	for _, c := range root.children {
		if weights[c] != cost {
			c_cost := 0
			for _, c_child := range vertices[c].children {
				c_cost += weights[c_child]
			}
			fixed := cost - c_cost
			return fixed
		}
	}
	weights[root.pid] = root.weight + len(root.children)*cost
	return -1
}

func balanceTree(programs []Program, root Program) int {
	vertices := make(map[string]Program)
	weights := make(map[string]int)

	for _, p := range programs {
		vertices[p.pid] = p
		weights[p.pid] = p.weight
	}
	return balance(root, vertices, weights)
}

func main() {
	dat, _ := ioutil.ReadFile("./7.txt")
	lines := strings.Split(string(dat), "\n")

	programs := make([]Program, len(lines))
	for idx, line := range lines {
		programs[idx] = parseLine(line)
	}

	source := findSource(programs)
	fmt.Printf("Part 1: %s\n", source.pid)

	fixed := balanceTree(programs, source)
	fmt.Printf("Part 2: %d\n", fixed)
}

File: Day1-9/7.py
import re
input_regex = re.compile('(.+) \((\d+)\)( -> )?(.+)?')


def parse_program(line):
    match = input_regex.match(line)
    pid = match.group(1)
    weight = int(match.group(2))
    if type(match.group(4)) == str:
        children = tuple(match.group(4).split(', '))
    else:
        children = ()
    return (pid, weight, children)


def find_source(programs):
    non_sources = set()
    vertices = set()
    for p in programs:
        pid, weight, children = p
        non_sources.update(children)
        vertices.update(children)
        vertices.add(pid)
    source = list(vertices - non_sources)[0]
    return source


def balance_tree(programs, root):
    vertices = {}
    weights = {}
    for p in programs:
        pid, weight, children = p
        vertices[pid] = p

    def _balance(root):
        pid, weight, children = vertices[root]
        if children:
            for c in children:
                fixed_weight = _balance(c)
                if fixed_weight:
                    return fixed_weight
            cost = min(weights[c] for c in children)
            for c in children:
                if weights[c] != cost:
                    c_pid, c_weight, c_children = vertices[c]
                    fixed = cost - sum(weights[c] for c in c_children)
                    return fixed
            weights[pid] = weight + len(children) * cost
        else:
            weights[pid] = weight
    return _balance(root)


if __name__ == '__main__':
    with open('7.txt') as f:
        lines = f.readlines()
    programs = [parse_program(p) for p in lines]
    source = find_source(programs)
    print("Part 1: {}".format(source))
    fixed_weight = balance_tree(programs, source)
    print("Part 2: {}".format(fixed_weight))
